articles/logging-with-winston/index.md

```diff
@@ -0,0 +1,576 @@
+
+Go ahead and install the [Express](/engineering-education/express/) library with `npm install express`.
+
+This is the structure of our small project:
+
+![Project Structure](/engineering-education/logging-with-winston/project-structure.png)
+
+- 📂Logs - will save the log files generated by the Winston file transport.
+- `app.js`🗄️ - will be our server app.
+- 📂Utils - will hold Winston `logger.js`🗄️ where we will add configurations such as Winston transport and formating.
+
+logger.js:
+
+```js
+const { createLogger, format, transports } = require('winston');
+require('winston-mongodb');
+
+module.exports = createLogger({
+transports:
+    new transports.File({
+    filename: './logs/example.log',
+    format:format.combine(
+        format.timestamp({format: 'MMM-DD-YYYY HH:mm:ss'}),
+        format.align(),
+        format.printf(info => `${info.level}: ${[info.timestamp]}: ${info.message}`),
+    )}),
+});
+```
+
+Will do some logging from an Express server that listens to port 3000.
+
+app.js:
+
+```js
+const express = require('express');
+
+// Require logger.js
+const logger = require('./utils/logger');
+const app = express();
+const port = 3000;
+const host = "localhost";
+
+// Dummy Express GET call
+app.get('/',(req,res) => {
+    res.send("Hello World!");
+    logger.info("Server Sent A Hello World!");
+})
+
+// Introduce error by using undefined variable 'y'
+app.get('/calc',(req,res) => {
+    const x = y + 10;
+    res.send(x.toString());
+})
+
+// Capture 500 errors
+app.use((err,req,res,next) => {
+res.status(500).send('Could not perform the calculation!');
+   logger.error(`${err.status || 500} - ${res.statusMessage} - ${err.message} - ${req.originalUrl} - ${req.method} - ${req.ip}`);
+})
+
+// Capture 404 erors
+app.use((req,res,next) => {
+    res.status(404).send("PAGE NOT FOUND");
+    logger.error(`400 || ${res.statusMessage} - ${req.originalUrl} - ${req.method} - ${req.ip}`);
+})
+// Run the server
+app.listen(port, () => {
+    console.log("Server started...");
+    logger.info(`Server started and running on http://${host}:${port}`)
+})
+```
+
+Run `node app.js` to start the server.
+
+From the above example:
+
+- Every time the server starts, Winston will record a log to the `example.log` file.
+
+When the server is running, accessing the following pages will create a log every time the link is invoked.
+
+- http://localhost:3000/ - the server will send a hello world message. We want Winston to capture that and record it in our log file.
+- http://localhost:3000/calc - we are trying to add variable `y` to variable `x`. In this case, variable `y` is not defined. This will generate an error, and we want Winston to capture such instances for us.
+- http://localhost:3000/hello - the server we have created has no such URL. We want Winston to inform us when a link that points to our IP address is accessed but can't be found; that is a `404` error.
+
+Go to the `example.log` file in the logs folder to view the log.
```

Comment:
Went to the routes above which loaded the expected result but there was no example.log file in the logs folder.